### PR TITLE
Upgrade libgit2 to 1.3.0, libssh2 to 1.10.2, mbedtls to 2.28 and libcurl to 7.81.0

### LINKED ISSUE #43249: LoadError: IOError: stream is closed or unusable with libgit2 1.3.0
Hello,

I'm updating libgit2 stack in Fedora from 1.1.0 to 1.3.0 and we compile Julia with USE_SYSTEM_LIBGIT2=1.

Sadly after libgit2 update, Julia tests (or some data generation for the?) fails with following stack trace:

```
Generating REPL precompile statements... 26/36ERROR: LoadError: IOError: stream is closed or unusable
Stacktrace:
  [1] check_open
    @ ./stream.jl:386 [inlined]
  [2] uv_write_async(s::Base.TTY, p::Ptr{UInt8}, n::UInt32)
    @ Base ./stream.jl:1018
  [3] uv_write(s::Base.TTY, p::Ptr{UInt8}, n::UInt32)
    @ Base ./stream.jl:981
  [4] unsafe_write(s::Base.TTY, p::Ptr{UInt8}, n::UInt32)
    @ Base ./stream.jl:1064
  [5] write
    @ ./strings/io.jl:244 [inlined]
  [6] write
    @ ./io.jl:639 [inlined]
  [7] (::Main.anonymous.var"#2#6"{Set{String}, String, Base.DevNull})(precompile_file::String, precompile_file_h::IOStream)
    @ Main.anonymous ~/build/BUILD/julia-1.7.0-rc3/contrib/generate_precompile.jl:326
  [8] mktemp(fn::Main.anonymous.var"#2#6"{Set{String}, String, Base.DevNull}, parent::String)
    @ Base.Filesystem ./file.jl:722
  [9] mktemp
    @ ./file.jl:720 [inlined]
 [10] generate_precompile_statements()
    @ Main.anonymous ~/build/BUILD/julia-1.7.0-rc3/contrib/generate_precompile.jl:270
 [11] top-level scope
    @ ~/build/BUILD/julia-1.7.0-rc3/contrib/generate_precompile.jl:416
 [12] eval(m::Module, e::Any)
    @ Core ./boot.jl:373
 [13] top-level scope
    @ ~/build/BUILD/julia-1.7.0-rc3/contrib/generate_precompile.jl:6
in expression starting at /builddir/build/BUILD/julia-1.7.0-rc3/contrib/generate_precompile.jl:3
```

While it does not seem to be an issue with libgit2 from the stacktrace, I've tried to build in almost the same environment (just with newer libgit2) and it does not fail on this error.

The issue is reproducible (I've tried to rebuild it ~5 times and it always fails like this).

here's the failing build log: https://kojipkgs.fedoraproject.org//work/tasks/2694/79372694/build.log
here's the successful build log: https://kojipkgs.fedoraproject.org//work/tasks/7535/79377535/build.log

---

I'm very happy to test out any patches or provide any more information, just let me know.

Thanks in advance!

### PR DESCRIPTION
**Pull request by nalimilan**
Goes with https://github.com/JuliaPackaging/Yggdrasil/pull/3962.

Closes #42209
Closes #43249

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Perhaps you need this patch? https://github.com/JuliaLang/julia/issues/43249#issuecomment-981147025

**Comment by KristofferC** [comment]
> Perhaps you need this patch?

That's already in here I think: https://github.com/JuliaLang/julia/pull/43250/files#diff-ba4b52130833cf4924e96d191202c63e7ddee40545d5c001cc927e77a392f762R233-R235

----

Closes https://github.com/JuliaLang/julia/pull/42209?

**Comment by DilumAluthge** [comment]
Hmmm. Not sure why tests are failing, then.

**Comment by DilumAluthge** [comment]
Ah. I guess the Ygg PR needs to be merged first.

File: deps/libgit2.mk [hunk 1]
```diff
@@ -40,28 +40,14 @@ $(LIBGIT2_SRC_PATH)/libgit2-agent-nonfatal.patch-applied: $(LIBGIT2_SRC_PATH)/so
...
 	echo 1 > $@
 
-# This can be removed once a release with https://github.com/libgit2/libgit2/pull/5685 lands
-$(LIBGIT2_SRC_PATH)/libgit2-mbedtls-incdir.patch-applied: $(LIBGIT2_SRC_PATH)/libgit2-agent-nonfatal.patch-applied
-	cd $(LIBGIT2_SRC_PATH) && \
-		patch -p1 -f < $(SRCDIR)/patches/libgit2-mbedtls-incdir.patch
-	echo 1 > $@
-
-$(LIBGIT2_SRC_PATH)/libgit2-hostkey.patch-applied: $(LIBGIT2_SRC_PATH)/libgit2-mbedtls-incdir.patch-applied
+$(LIBGIT2_SRC_PATH)/libgit2-hostkey.patch-applied:
```
Comment (t-bltg) [inline deps/libgit2.mk:43]:
```suggestion
$(LIBGIT2_SRC_PATH)/libgit2-hostkey.patch-applied: $(LIBGIT2_SRC_PATH)/libgit2-agent-nonfatal.patch-applied
```

Comment (nalimilan) [reply deps/libgit2.mk:43]: That file is no longer created. Why do you want to revert this?

Comment (t-bltg) [reply deps/libgit2.mk:43]: This target needs to depends on `$(LIBGIT2_SRC_PATH)/source-extracted` which is a dependecy of `$(LIBGIT2_SRC_PATH)/libgit2-agent-nonfatal.patch-applied` else you run into concurrent issues, trying to patch a source file before it is extracted.

**Comment by mkitti** [comment]
Just reporting in from https://github.com/conda-forge/julia-feedstock. With PR https://github.com/conda-forge/julia-feedstock/pull/157, we deployed Julia 1.7.1 with LibGit2 1.3.

**Comment by ViralBShah** [comment]
Note that this will need Mbedtls to also update to 0.28.

**Comment by ViralBShah** [comment]
mbedtls, libssh2 and libgit2 should probably all be upgraded simultaneously, since libssh2 and libgit2 are hard linked to mbedtls 2.28

File: deps/libgit2.mk (hunk 1 above)
Comment (ViralBShah) [reply deps/libgit2.mk]: IS this resolved?

**Review by ViralBShah — APPROVED** [review]

File: deps/libgit2.mk (hunk 1 above)
Comment (t-bltg) [reply deps/libgit2.mk]: Looks like it is.

**Comment by mkitti** [comment]
Looks like libssh2 needs to be updated on this branch:
https://github.com/JuliaLang/julia/blob/nl/libgit2-1.3.0/deps/libssh2.version

Currently, it says 1.10.0. https://github.com/JuliaLang/julia/blob/65a56215733c0da8d028eff224bc79a57552ccaf/deps/libssh2.version#L1-L2

Shouldn't it be 1.10.2 https://github.com/JuliaBinaryWrappers/LibSSH2_jll.jl/releases/tag/LibSSH2-v1.10.2%2B0 ?

**Comment by nalimilan** [comment]
> The build is still looking for libmbedtls.so.13

Yeah, but AFAICT this is due to libgit2.so (and then libssh2.so and libcurl.so) linking to libmbedtls.so.13. I guess they need to be rebuilt in Yggdrasil? The build works with `USE_BINARYBUILDER_*=0` for these three deps.

> Shouldn't it be 1.10.2 https://github.com/JuliaBinaryWrappers/LibSSH2_jll.jl/releases/tag/LibSSH2-v1.10.2%2B0 ?

This is one of the things I wondered. Since the variable says "branch", I assumed it has to refer to an existing branch in the repo? 1.10.2 doesn't actually exist upstream.

**Comment by mkitti** [comment]
> This is one of the things I wondered. Since the variable says "branch", I assumed it has to refer to an existing branch in the repo? 1.10.2 doesn't actually exist upstream.

This should be the same as this:
https://github.com/JuliaPackaging/Yggdrasil/blob/dd0bfdabbaef3493841f54bce34539b270ebab8c/L/LibSSH2/build_tarballs.jl#L7-L11

**Comment by nalimilan** [comment]
OK, turns out I hadn't updated JLLs for all of these! But we'll need to update Curl for build to work: https://github.com/JuliaPackaging/Yggdrasil/pull/4200

**Comment by nalimilan** [comment]
Windows failures seem to be unrelated as they are also on e.g. https://build.julialang.org/#/builders/63/builds/2154/steps/5/logs/stdio.

**Comment by giordano** [comment]
I don't think this should be backported, or at least we need to think about how to handle the backport: this will break lots of JLLs and related packages

**Comment by mkitti** [comment]
We may need to consider what long term support means to us and what resources we have provide to provide long term support in that case.

I suppose the ideal scenario is that we backport the security fixes in a way does not break the ABI of these dependencies.

Going forward, we should also consider if the original reason we chose these dependencies over alternatives is still applicable. For example, some distributions are considering moving from `libssh2` to `libssh` because of upstream support issues similar to what we have experienced.
1. https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/WE3UHS5W4T7XTPOEPMUFJE433XFNO5MC/
2. https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=897950

Similar questions about libgit2 using libssh2 were previously raised:
1. https://github.com/JuliaLang/Pkg.jl/issues/2679 (by @StefanKarpinski)
2. https://github.com/libgit2/libgit2/pull/5253 (however, a libssh port of libgit2 is being considered)

The other complicating matter in this is also mbedtls.

Their release considerations and policy about API changes do not seem conducive to Julia's long term support model and binary distribution strategy:

> API changes

> Some fields of mbedtls_ssl_session and mbedtls_ssl_config are in a
different order. This only affects applications that define such
structures directly or serialize them.

https://github.com/ARMmbed/mbedtls/releases/tag/v2.28.0

**Comment by ViralBShah** [comment]
While we need to think carefully about all this going forward, I do think we need to backport (something) to 1.6 and 1.7, run pkgeval and see how bad the breakage is.

**Comment by mkitti** [comment]
It sounds like this is going to be a lot of work for @giordano . Let me know if I can help in anyway. The only scenario I can imagine is an effective hard fork of 1.6.

**Comment by giordano** [comment]
There are over 1400 packages [depending on libssh2](https://juliahub.com/ui/Packages/LibSSH2_jll/K6mup/1.10.2+0?page=2), to give the magnitude of the breakage

**Comment by ViralBShah** [comment]
But very few direct ones, which makes me feel like it may not be as hard as it looks.

**Comment by giordano** [comment]
[`NetCDF_jll`](https://juliahub.com/ui/Packages/NetCDF_jll/cIvCy/page=2) is one that I know for sure links directly to mbedtls:
```console
% readelf -d libnetcdf.so | head

Dynamic section at offset 0x1a0658 contains 35 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libnghttp2.so.14]
 0x0000000000000001 (NEEDED)             Shared library: [libssh2.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libmbedtls.so.13]
 0x0000000000000001 (NEEDED)             Shared library: [libmbedx509.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libmbedcrypto.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libhdf5_hl.so.200]
```

**Comment by ViralBShah** [comment]
Is the issue that the ABI breakage will break everything that links to mbedtls?

**Comment by nalimilan** [comment]
It's probably better (both in terms of avoiding breakage and in terms of required work) to only backport patches that are needed to fix security issues. Note that Linux distributions have to do the same thing (notably Ubuntu LTS and CentOS, but that also applies to others over their shorter lifetimes), so with some luck we could just reuse the patches they backport. That's maybe something to keep in mind for the next LTS: using the same versions as a major LTS distribution would ensure that we can do that easily.

**Comment by nalimilan** [comment]
Merge?

**Comment by mkitti** [comment]
Why is buildbot/tester_win32 pending?

**Comment by ViralBShah** [comment]
We have the CI straight flush. Haven't seen that in a while. Good to merge?

**Comment by ViralBShah** [comment]
I've removed the backport labels, since we may want to do a very limited backport and not this whole PR.

**Comment by nalimilan** [comment]
64-bit Linux CI failed, but that doesn't seem related.

**Comment by Seelengrab** [comment]
We may want to backport the libssh2 version bump to LTS: https://cve.circl.lu/cve/CVE-2019-17498

**Comment by giordano** [comment]
That's not really doable without breaking [some thousands of packages](https://juliahub.com/ui/Packages/MbedTLS_jll/u5NEn/2.28.0+0?page=2) indirectly depending on mbedtls along the way

**Comment by Seelengrab** [comment]
I'm not sure I understand - is libssh2 deeply tied to mbedtls? I think I'm missing some context here, since just some messages up you mentioned that libssh2 may be possible: 

> I could understand backporting stuff like libgit2 and libssh2 only which have a stable ABI (and I'd still be wary of doing that blindly), but mbedtls is a serious mess.

Though I of course understand that the impact of that CVE could be considered low enough to not be an issues. I definitely don't want to imply that it has to be done!

**Pull request merged**